[최용석] Step4 PR 리뷰 요청드립니다.

badaelephant opened the pull request:
안녕하세요 장민제 리뷰어님, 

경로조회 관련 기능 리뷰 요청드립니다. 

이번에도 리뷰 잘 부탁드려요 ㅎㅎ

Commit:
refactor: 피드백 반영
1. 즐겨찾기 비교로직 수정
2. 즐겨찾기 통합테스트 displayname 수정

Commit:
test: 경로 조회 인수테스트 수정 - 요금 조회 부분 추가

Commit:
test: 경로조회 서비스 테스트 수정

Commit:
feat: 경로조회 컨트롤러 기능 수정
LoginMember 관련 인증 로직 수정

Commit:
feat: 경로조회 인터페이스에 로그인 멤버 경로조회 메서드 추가

Commit:
feat: 경로조회 서비스 기능수정

Commit:
test: 경로조회 테스트 코드에 요금조회 테스트코드 추가

Commit:
test: 요금 단위테스트 추가

Commit:
feat: 요금 도메인 기능 추가

Commit:
feat: 경로 도메인 추가

Commit:
feat: 노선에 추가요금 관련된 로직 수정 - 노선 엣지 도메인 추가

Commit:
test: 할인 테스트 코드 추가

Commit:
feat: 할인 기능 추가

Commit:
feat: 경로조회 기능 객체 수정

Commit:
feat: 마이너 변경

Commit:
docs: step4 리드미 수정

minzzang commented on README.md:
```diff
@@ -83,4 +83,21 @@ Feature: 지하철 구간 관련 기능
+
+### 요구사항
+
+- [x] 경로 조회 시 거리 기준 요금 정보 포함하기
```
요구사항 정리 👍

badaelephant replied:
감사합니다.

minzzang commented on src/main/java/nextstep/subway/path/domain/DijkstraShortestPathFinder.java:
```diff
@@ -5,24 +5,36 @@
+    }
+
+    @Override
+    public Path findPathByLoginMember(Station source, Station target, Integer age) {
```
길 찾기 기능과 요금을 계산하는 부분이 같이 있는 것 같은데 PathFinder에서는 길을 찾는 책임만 가지고 요금 계산 로직은 분리를 하는게 좋을 것 같아요! 😄

badaelephant replied:
이부분 말씀하신대로 수정하는게 좋다는 것을 테스트 코드에서 느꼈습니다. 테스트 코드가 기능별로 분화가 되니까, 확실히 이 테스트 코드가 어떤것을 검증하고 싶은지를 명확하게 구별할 수 있었습니다. 또한, 굳이 로그인 멤버를 경로조회에 인자로 넘겨줄 필요가 없는데 이전에는 넘겨주고 있었습니다. 로그인멤버를 딱 요금에서만 활용하니까 불필요하게 들고다닐 필요가 없어서 훨씬 깔끔해졌습니다. 

그동안 역할분리를 많이 들었지만, 뭔가 이번에는 테스트 코드와 로직에서 기능이 너무 섞여있어서 불편했었는데 리뷰 듣고 수정해서 깔끔해지니 필요성을 확실히 느꼈습니다. 감사해요^^

minzzang commented on src/main/java/nextstep/subway/path/domain/Fare.java:
```diff
@@ -0,0 +1,67 @@
+    public static Fare ofDiscount(Distance distance, int surCharge, Discount discount){
+        return new Fare(distance.value(), surCharge, discount);
+    }
+    private void addByDistance(int distance){
```
거리별 요금 정책을 계산하는 로직인데 요금과 요금 정책을 분리해보면 어떨까요~?

badaelephant replied:
이부분도 위와 동일하게 나누니까 훨씬 깔끔하더라구요. 

다만, 수정하면서, 거리와 할인률 요금정책(DistanceFareStrategy, DiscountFareStrategy)의 경우는 역할이 명확한 느낌인데, 경로 구간의 최고 추가요금을 구하는 것은 이미 일급컬렉션인 SectionEdges에서 하고 있어서, LineSurChargeFareStrategy와 같은 별도의 전략을 만드는 것이 아니라, 그냥 가져온 SectionEdges를 바로 활용했습니다. 

이부분도 크게 이상하지 않나요? 더 좋은 방법이 있으면 알려주시면 감사하겠습니다.

minzzang commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -26,6 +27,10 @@ public boolean supportsParameter(MethodParameter parameter) {
         String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        AuthenticationPrincipal authenticationPrincipal = parameter.getParameterAnnotation(AuthenticationPrincipal.class);
+        if(!authenticationPrincipal.required()){
+            return LoginMember.GUEST;
```
guest분리 👍

minzzang commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -1,34 +1,52 @@
+    public static Path ofAge(GraphPath graphPath, Integer age) {
+        Distance distance = Distance.from((int)graphPath.getWeight());
+        int maxSurCharge = getMaxSurCharge(graphPath.getEdgeList());
+        Discount discount = Discount.getDiscountByAge(age.intValue());
```
Path에서 구체적으로 요금과 관련된 정책을 아는 것보단 정책을 추상화 시키고 요금에 대한 값만 반환 받는 식으로 변경하면 좋을 것 같아요! 현재는 여러 군데에서 요금을 계산하고 있는데 그 부분을 객체지향적으로 어떻게 추상화하고 어디에 책임을 부여할지 고민을 해보면 좋을 것 같아요!

badaelephant replied:
이번 리뷰를 통해 이부분 많이 느끼고 공감했습니다. 감사합니다. 수정내용 보시고 아직 기능적으로 더 명확히 분할 해야할 부분이 있다면 말씀해주세요 ㅎㅎ

minzzang reviewed (changes requested):
안녕하세요 용석님~!

요금 조회 미션 잘 작성해주셨네요! :+1:

요금 별 정책 로직을 좀 더 추상화해보면 좋을 것 같아 코멘트와 질문을 남겼는데 확인 부탁드립니다. :pray:

오늘 하루도 화이팅입니다. :muscle:

Commit:
refactor: 피드백 반영
1. 역할분리 : 경로조회와 요금 관련 기능을 분리
- 요금 기능의 경우도, 요금을 계산하는 기능은 FareStrategy 로 분리.
2. 일급컬렉션 SectionEdges
3. 패키지 구조 정리(요금, 경로)
4. 정적팩토리 메소드 적용
5. 역할분화에따라 테스트 위치 변경 - 경로는 경로만, 요금은 요금만 테스트

minzzang commented on src/main/java/nextstep/subway/path/application/PathFinderService.java:
```diff
@@ -26,11 +31,16 @@ public PathFinderService(SectionRepository sectionRepository, StationService sta
+        Path path = Path.of(graphPath, Fare.of(fareStrategy));
+        return PathResponse.from(path);
+    }
+    private FareStrategy createFareStrategy(GraphPath graphPath, LoginMember loginMember){
```
요금 정책 분리 👍

minzzang commented on src/main/java/nextstep/subway/path/domain/path/SectionEdges.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Collections;
+import java.util.List;
+
+public class SectionEdges {
```
일급 컬렉션 분리 👍  
DistanceFareStrategy같이 이 클래스도 요금 정책을 계산하는 책임을 가지고 있는데 네이밍을 비슷하게 맞춰보면 더 좋을것 같아요! 😄

minzzang commented on src/main/java/nextstep/subway/path/domain/fare/FareStrategy.java:
```diff
@@ -0,0 +1,33 @@
+import nextstep.subway.line.domain.Distance;
+import nextstep.subway.path.domain.path.SectionEdges;
+
+public class FareStrategy {
```
요금 전략을 아주 잘 분리해주셨네요! 가독성이 엄청 좋아졌습니다!

지금 요구사항에서는 군더더기 없는 구조이지만 나중에 전략이 다양해져서 분리가 필요할 경우에는 디자인 패턴에 책임 연쇄 패턴, 컴포짓 패턴을 적용해봐도 좋을 것 같아요! 😄

minzzang reviewed (approved):
안녕하세요 용석님~!

피드백 잘 반영해주셨네요!! :+1:
책임 분리를 잘 해주셔서 코드를 쉽게 읽을 수 있었습니다! :100:

이번 미션을 통해 인수 테스트와 단위 테스트 작성, 객체지향 설계에 도움이 되셨으면 좋겠습니다!
용석님의 미션에 임하는 자세와 열정 덕분에 저 역시 리뷰를 하면서 많이 고민하고 배우게 된 것 같습니다.

저와의 미션은 여기서 끝이지만 다른 미션 진행하면서도 궁금한 점이나 어려운 부분 있으면 편하게 dm 해주세요 :smile:

인수테스트 기반 TDD 미션 수행하시느라 정말 고생많으셨습니다.
이제 2주 정도 남았는데 남은 미션들도 모두 완주하셔서 좋은 결과있길 바라겠습니다! :smiley:

badaelephant commented:
> 안녕하세요 용석님~!
> 
> 피드백 잘 반영해주셨네요!! 👍 책임 분리를 잘 해주셔서 코드를 쉽게 읽을 수 있었습니다! 💯
> 
> 이번 미션을 통해 인수 테스트와 단위 테스트 작성, 객체지향 설계에 도움이 되셨으면 좋겠습니다! 용석님의 미션에 임하는 자세와 열정 덕분에 저 역시 리뷰를 하면서 많이 고민하고 배우게 된 것 같습니다.
> 
> 저와의 미션은 여기서 끝이지만 다른 미션 진행하면서도 궁금한 점이나 어려운 부분 있으면 편하게 dm 해주세요 😄
> 
> 인수테스트 기반 TDD 미션 수행하시느라 정말 고생많으셨습니다. 이제 2주 정도 남았는데 남은 미션들도 모두 완주하셔서 좋은 결과있길 바라겠습니다! 😃

안녕하세요 민제님 ^-^ 이번 리뷰 정말 잘 해주셔서 많이 배울 수 있었습니다. 

좋은 하루 되시고 앞으로 좋은 기회로 꼭 만나뵐 수 있으면 좋겠네요 ~ 행복한 연말 되세요
